### PR TITLE
Revamp CI/CD

### LINKED ISSUE #165: Generate coverage report for ESBMC
## Summary

We have lots of regressions, unit/fuzzing tests and yet do not know which parts of the code are well tested and which are not. Since ESBMC is now OSS and uses CMake we could use external tools for it. I propose using [codecov](https://github.com/marketplace/codecov/) which is free for OSS.

## Tasks

- Generate the coverage report using CMake, we could integrate [this module](https://github.com/RWTH-HPC/CMake-codecov)
- Use the codecov cmake module inside the CI and add the badge to the top of the README.

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
Some major changes in the CI infrastructure

- Removed Jenkinsfile.
- ~~Removed build action.~~ I reactivated it to build with all supported solvers in Linux/Darwin.
- Renamed some of the actions names.
- Added linter action with: code-style, super-linter, cmakelint.
- Added clang-tidy support in CMake together with a github action.
- Added coverage report generation with gcov and lcov.
- Added Travis support with builds using g++, clang, each sanitizer and osx 10.14. Travis has a limit of 50min so I disabled CVC4 since it can take up to 20min if the cache is not hit.
- The Linux g++ build on Travis also submits the coverage report into Codecov. Should close https://github.com/esbmc/esbmc/issues/165

Notes: 
- Badges will be added in a later PR.
- We should open a discussion about what checks clang-tidy should do.
- Sanitizers and linters are throwing errors but are not marking the job as failed. ESBMC contains errors that go way beyond the scope of this PR.

### CODE REVIEW

File: regression/CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,14 @@
...
 enable_testing()
 
+# ctest_empty_binary_directory("${CTEST_BINARY_DIRECTORY}")
+
+#set(CTEST_SITE "${CMAKE_SYSTEM_NAME}")
+#set(CTEST_BUILD_NAME "experimental build")
+
+#set(CTEST_CMAKE_GENERATOR "Ninja")
+#set(CTEST_COVERAGE_COMMAND "gcov")
+
```
Comment (mikhailramalho) [inline regression/CMakeLists.txt]: Remove comments

**Review by mikhailramalho — APPROVED** [review]
LGTM.

Are we going to run travis in Manchester?

**Comment by rafaelsamenezes** [comment]
> Are we going to run travis in Manchester?

No, Travis is free for open-source projects.

File: .github/workflows/linter.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      - name: Lint Code Base
+        uses: docker://github/super-linter:v2.1.1
+        env:
+          VALIDATE_ALL_CODEBASE: true
+          VALIDATE_PYTHON: true
+          VALIDATE_BASH: true
+          VALIDATE_MD: true
+          VALIDATE_YAML: true
+          VALIDATE_XML: true
+          DISABLE_ERRORS: true # TODO: Eventually this will be removed
```
Comment (feliperodri) [inline .github/workflows/linter.yml:25]: I'd suggest to remove the comment and create a GitHub issue for this.

File: .github/workflows/linter.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      run: tar xf clang+llvm-9.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz && mv clang+llvm-9.0.0-x86_64-linux-gnu-ubuntu-18.04 clang9
+    - name: Run clang-format
+      run: find src -iname *.h -o -iname *.c -o -iname *.cpp -o -iname *.hpp -iname *.hh | grep -v src/ansi-c/cpp | xargs ./clang9/bin/clang-format -style=file -i -fallback-style=none
+    - name: Throws error if changes were made
+      run: git diff --exit-code
+
+  cmake-lint:  
+    name: Check CMake modules
+    runs-on: ubuntu-latest
+    continue-on-error: true # TODO: Eventually this will be removed
```
Comment (feliperodri) [inline .github/workflows/linter.yml:45]: I'd suggest to remove the comment and create a GitHub issue for this.

File: regression/CMakeLists.txt (hunk 1 above)
Comment (rafaelsamenezes) [reply regression/CMakeLists.txt]: Fixed

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    - os: linux
+      compiler: "clang"
+      env: ENABLE_SANITIZER=1
+           SANITIZER="UBSAN"
+    - os: osx
+      osx_image: xcode10.2
+      compiler: "clang"
+    
+  allow_failures:
+    # TODO: Identify, fix, and remove each build mode from failure
```
Comment (feliperodri) [inline .travis.yml:79]: comment -> GitHub issue.

**Review by feliperodri — APPROVED** [review]

File: .github/workflows/linter.yml (hunk 2 above)
Comment (rafaelsamenezes) [reply .github/workflows/linter.yml:25]: I think it is best to create the issue and mantain the comment to help to locate were the change is needed.

**Review by lucasccordeiro — APPROVED** [review]
LGTM.

**Pull request merged**